From 503bb3941a945eef93b7a93e2fb14c06a77545b3 Mon Sep 17 00:00:00 2001
From: t <tristan75015@gmail.com>
Date: Wed, 21 Oct 2026 03:33:11 +0200
Subject: [PATCH] Behaviours: functions of batches and of a point detected
 independently

As for the material properties, the functions of batches of points exist if
and only if the library exports `<name>_generic_parallel_api_version`
(`BehaviourDescription::batch_api_version`, read by
`isGenericParallelBehaviour`), and the functions of a point if and only if
the library exports `Generic` as interface (or no interface). Both are loaded
when a library of the `generic-parallel` interface also contains the
functions of the `generic` interface; `checkBatchInterface` applies this
rule for both kinds of functions.
---
 include/MGIS/Behaviour/Behaviour.hxx          | 19 ++++++++------
 .../MGIS/Behaviour/BehaviourDescription.hxx   | 18 ++++++++++---
 include/MGIS/LibrariesManager.hxx             | 18 +++++++------
 src/Behaviour.cxx                             | 23 +++++++++++++----
 src/BehaviourDescription.cxx                  |  6 +++--
 src/Integrate.cxx                             |  3 ++-
 src/LibrariesManager.cxx                      | 25 ++++++++++++-------
 7 files changed, 76 insertions(+), 36 deletions(-)

diff --git a/include/MGIS/Behaviour/Behaviour.hxx b/include/MGIS/Behaviour/Behaviour.hxx
index 8019058..d96daba 100644
--- a/include/MGIS/Behaviour/Behaviour.hxx
+++ b/include/MGIS/Behaviour/Behaviour.hxx
@@ -115,9 +115,10 @@ namespace mgis::behaviour {
     BehaviourFctPtr b = nullptr;
     /*!
      * \brief pointer to the function integrating the behaviour on a batch of
-     * integration points (`generic-parallel` interface). In this case, the
-     * pointer `b` and the pointers to the rotation functions of a point are
-     * null.
+     * integration points (`generic-parallel` interface). The pointer `b`
+     * and the pointers to the rotation functions of a point are then null,
+     * unless the library also provides the functions of the `generic`
+     * interface (it then exports `Generic` as interface).
      */
     BatchBehaviourFctPtr batch_b = nullptr;
     /*!
@@ -223,7 +224,8 @@ namespace mgis::behaviour {
    * overload taking `FiniteStrainBehaviourOptions`.
    * \note for a behaviour generated by the `generic-parallel` interface, the
    * functions processing batches of integration points are loaded (members
-   * `batch_*`) and the pointers to the functions of a point are null (see
+   * `batch_*`) and the pointers to the functions of a point are null, unless
+   * the library also provides the functions of the `generic` interface (see
    * also `loadBehaviourDescription`).
    */
   MGIS_EXPORT MGIS_NODISCARD std::optional<Behaviour> load(
@@ -244,7 +246,8 @@ namespace mgis::behaviour {
    * meaningful here
    * \note for a behaviour generated by the `generic-parallel` interface, the
    * functions processing batches of integration points are loaded (members
-   * `batch_*`) and the pointers to the functions of a point are null (see
+   * `batch_*`) and the pointers to the functions of a point are null, unless
+   * the library also provides the functions of the `generic` interface (see
    * also `loadBehaviourDescription`).
    */
   MGIS_EXPORT MGIS_NODISCARD std::optional<Behaviour> load(
@@ -264,7 +267,8 @@ namespace mgis::behaviour {
    * overload taking `FiniteStrainBehaviourOptions`.
    * \note for a behaviour generated by the `generic-parallel` interface, the
    * functions processing batches of integration points are loaded (members
-   * `batch_*`) and the pointers to the functions of a point are null (see
+   * `batch_*`) and the pointers to the functions of a point are null, unless
+   * the library also provides the functions of the `generic` interface (see
    * also `loadBehaviourDescription`).
    */
   MGIS_EXPORT Behaviour load(const std::string &,
@@ -282,7 +286,8 @@ namespace mgis::behaviour {
    * meaningful here
    * \note for a behaviour generated by the `generic-parallel` interface, the
    * functions processing batches of integration points are loaded (members
-   * `batch_*`) and the pointers to the functions of a point are null (see
+   * `batch_*`) and the pointers to the functions of a point are null, unless
+   * the library also provides the functions of the `generic` interface (see
    * also `loadBehaviourDescription`).
    */
   MGIS_EXPORT Behaviour load(const FiniteStrainBehaviourOptions &,
diff --git a/include/MGIS/Behaviour/BehaviourDescription.hxx b/include/MGIS/Behaviour/BehaviourDescription.hxx
index 522c3a46..8545d8d 100644
--- a/include/MGIS/Behaviour/BehaviourDescription.hxx
+++ b/include/MGIS/Behaviour/BehaviourDescription.hxx
@@ -63,9 +63,18 @@ namespace mgis::behaviour {
     std::string unit_system;
     /*!
      * \brief interface of `MFront` used to generate the behaviour: `Generic`
-     * (or empty for old libraries) or `GenericParallel`
+     * (or empty for old libraries) or `GenericParallel`. The functions of a
+     * point are available if and only if the interface is `Generic` (or
+     * empty): a library of the `generic-parallel` interface also containing
+     * the functions of the `generic` interface exports `Generic`.
      */
     std::string interface;
+    /*!
+     * \brief version of the data structures of the functions of batches of
+     * points (`generic-parallel` interface), or `0` if the library does not
+     * provide them (no `<name>_generic_parallel_api_version` symbol)
+     */
+    unsigned short batch_api_version = 0;
     /*!
      * \brief backend of a library generated by the `generic-parallel`
      * interface (`cuda`, `hip`, `stlpar`, `kokkos`, `raja`, `sycl`, `openmp`
@@ -167,9 +176,10 @@ namespace mgis::behaviour {
                                             const std::string &,
                                             const Hypothesis);
   /*!
-   * \return true if the behaviour has been generated by the
-   * `generic-parallel` interface of `MFront`, i.e. integrates batches of
-   * integration points
+   * \return true if the library of the behaviour provides the functions of
+   * batches of integration points (`generic-parallel` interface of `MFront`),
+   * i.e. if the version of their data structures has been read
+   * (`batch_api_version` member)
    * \param[in] d: behaviour description
    */
   MGIS_EXPORT bool isGenericParallelBehaviour(
diff --git a/include/MGIS/LibrariesManager.hxx b/include/MGIS/LibrariesManager.hxx
index 8f94489..1de983d 100644
--- a/include/MGIS/LibrariesManager.hxx
+++ b/include/MGIS/LibrariesManager.hxx
@@ -995,17 +995,19 @@ namespace mgis {
     //! \brief constructor
     LibrariesManager();
     /*!
-     * \brief check that an entry point has (`batch` true) or has not been
-     * generated by the `generic-parallel` interface, and throw an exception
-     * otherwise. The functions of this interface have the names of the
-     * functions of the `generic` interface, but process batches of points.
-     * The numerical type of the entry point is also checked (see
-     * `checkRealType`).
+     * \brief check that an entry point provides the functions of batches of
+     * points (`batch` true) or the functions of a point (`batch` false), and
+     * throw an exception otherwise. The functions of batches of points exist
+     * if and only if the `<name>_generic_parallel_api_version` symbol is
+     * exported, the functions of a point if and only if the library exports
+     * `Generic` as interface (or no interface, for old libraries): a library
+     * of the `generic-parallel` interface may provide both. The numerical
+     * type of the entry point is also checked (see `checkRealType`).
      * \param[in] m: calling method
      * \param[in] l: library name
      * \param[in] n: entry point name
-     * \param[in] batch: if true, the entry point must have been generated by
-     * the `generic-parallel` interface. Otherwise, it must not.
+     * \param[in] batch: if true, the functions of batches of points are
+     * required, otherwise the functions of a point
      */
     void checkBatchInterface(const char *const,
                              const std::string &,
diff --git a/src/Behaviour.cxx b/src/Behaviour.cxx
index 83137bc..8f2acdc 100644
--- a/src/Behaviour.cxx
+++ b/src/Behaviour.cxx
@@ -80,11 +80,16 @@ namespace mgis::behaviour {
     //
     auto d = Behaviour{};
     loadBehaviourDescription(d, l, b, h);
+    // the functions of batches of points and the functions of a point are
+    // detected independently: a library of the `generic-parallel` interface
+    // may provide both
     const auto batch = isGenericParallelBehaviour(d);
+    const auto pointwise = (d.interface.empty()) || (d.interface == "Generic");
     if (batch) {
       d.batch_b = lm.getBatchBehaviour(l, b, h);
       d.batch_report = lm.getBatchBehaviourReportFunction(l, b);
-    } else {
+    }
+    if (pointwise) {
       d.b = lm.getBehaviour(l, b, h);
     }
     //
@@ -96,7 +101,8 @@ namespace mgis::behaviour {
       BehaviourInitializeFunction ifct;
       if (batch) {
         ifct.batch_f = lm.getBatchBehaviourInitializeFunction(l, b, i, h);
-      } else {
+      }
+      if (pointwise) {
         ifct.f = lm.getBehaviourInitializeFunction(l, b, i, h);
       }
       ifct.inputs = getBehaviourInitializeFunctionInputs(l, b, i, h);
@@ -107,7 +113,8 @@ namespace mgis::behaviour {
       BehaviourPostProcessing pfct;
       if (batch) {
         pfct.batch_f = lm.getBatchBehaviourPostProcessing(l, b, i, h);
-      } else {
+      }
+      if (pointwise) {
         pfct.f = lm.getBehaviourPostProcessing(l, b, i, h);
       }
       pfct.outputs = getBehaviourPostProcessingOutputs(l, b, i, h);
@@ -136,7 +143,10 @@ namespace mgis::behaviour {
           lm.getBatchRotateBehaviourThermodynamicForcesFunction(l, b, h);
       d.batch_rotate_tangent_operator_blocks_ptr =
           lm.getBatchRotateBehaviourTangentOperatorBlocksFunction(l, b, h);
-    } else if (d.symmetry == Behaviour::ORTHOTROPIC) {
+    }
+    // rotation functions of a point, if the library provides the functions
+    // of a point
+    if ((d.symmetry == Behaviour::ORTHOTROPIC) && (d.b != nullptr)) {
       auto &lm = mgis::LibrariesManager::get();
       d.rotate_gradients_ptr = lm.getRotateBehaviourGradientsFunction(l, b, h);
       d.rotate_array_of_gradients_ptr =
@@ -232,7 +242,10 @@ namespace mgis::behaviour {
       d.batch_rotate_tangent_operator_blocks_ptr =
           lm.getBatchRotateBehaviourTangentOperatorBlocksFunction(
               l, b, h, o.tangent_operator);
-    } else if (d.symmetry == Behaviour::ORTHOTROPIC) {
+    }
+    // rotation functions of a point, if the library provides the functions
+    // of a point
+    if ((d.symmetry == Behaviour::ORTHOTROPIC) && (d.b != nullptr)) {
       auto &lm = mgis::LibrariesManager::get();
       d.rotate_gradients_ptr = lm.getRotateBehaviourGradientsFunction(l, b, h);
       d.rotate_array_of_gradients_ptr =
diff --git a/src/BehaviourDescription.cxx b/src/BehaviourDescription.cxx
index 6b05915..91bb593 100644
--- a/src/BehaviourDescription.cxx
+++ b/src/BehaviourDescription.cxx
@@ -123,7 +123,7 @@ namespace mgis::behaviour {
   BehaviourDescription::~BehaviourDescription() = default;
 
   bool isGenericParallelBehaviour(const BehaviourDescription &d) noexcept {
-    return d.interface == "GenericParallel";
+    return d.batch_api_version != 0;
   }  // end of isGenericParallelBehaviour
 
   bool isMemoryLocationAccessible(const BehaviourDescription &d,
@@ -172,7 +172,8 @@ namespace mgis::behaviour {
       raise(msg);
     }
     d.interface = lm.getInterface(l, b);
-    if (isGenericParallelBehaviour(d)) {
+    // functions of batches of points, if their version is exported
+    if (lm.hasGenericParallelFunctions(l, b)) {
       constexpr auto gp_api_version = MGIS_BV_GENERIC_PARALLEL_API_VERSION +
                                       MGIS_REAL_TYPE_API_VERSION_OFFSET;
       const auto v = lm.getGenericParallelAPIVersion(l, b);
@@ -181,6 +182,7 @@ namespace mgis::behaviour {
                "generic-parallel interface (the behaviour uses version " +
                    std::to_string(v) + ", mgis uses version " +
                    std::to_string(gp_api_version) + ")");
+      d.batch_api_version = v;
       d.backend = lm.getGenericParallelBackend(l, b);
       d.memory_space = lm.getGenericParallelMemorySpace(l, b);
       raise_if(!isValidMemorySpace(d.memory_space),
diff --git a/src/Integrate.cxx b/src/Integrate.cxx
index c49c924..d6f9a67 100644
--- a/src/Integrate.cxx
+++ b/src/Integrate.cxx
@@ -41,7 +41,8 @@ namespace mgis::behaviour::internals {
                 ": no function of an integration point (the behaviour has "
                 "been generated by the generic-parallel interface, whose "
                 "functions integrate batches of integration points: see the "
-                "functions of the material data managers)");
+                "functions of the material data managers, or the "
+                "generic_functions option of this interface)");
   }  // end of raiseMissingFunctionOfAPoint
 
   static void allocate(MaterialDataManager& m,
diff --git a/src/LibrariesManager.cxx b/src/LibrariesManager.cxx
index 045fa40..52ea41f 100644
--- a/src/LibrariesManager.cxx
+++ b/src/LibrariesManager.cxx
@@ -962,19 +962,26 @@ namespace mgis {
                                              const std::string &n,
                                              const bool batch) {
     this->checkRealType(m, l, n);
-    const auto generic_parallel = this->getInterface(l, n) == "GenericParallel";
-    if (generic_parallel == batch) {
+    // the functions of batches of points exist if and only if their version
+    // is exported, and the functions of a point if and only if the library
+    // exports `Generic` as interface (a library of the `generic-parallel`
+    // interface may contain both)
+    if (batch) {
+      if (!this->hasGenericParallelFunctions(l, n)) {
+        mgis::raise(std::string{m} + ": entry point '" + n + "' in library '" +
+                    l +
+                    "' has not been generated by the generic-parallel "
+                    "interface");
+      }
       return;
     }
-    if (batch) {
+    const auto i = this->getInterface(l, n);
+    if ((!i.empty()) && (i != "Generic")) {
       mgis::raise(std::string{m} + ": entry point '" + n + "' in library '" +
-                  l +
-                  "' has not been generated by the generic-parallel "
-                  "interface");
+                  l + "' has been generated by the '" + i +
+                  "' interface, which provides no function of a point: the "
+                  "functions of batches of points must be used");
     }
-    mgis::raise(std::string{m} + ": entry point '" + n + "' in library '" + l +
-                "' has been generated by the generic-parallel interface: "
-                "the functions of batches of points must be used");
   }  // end of checkBatchInterface
 
   unsigned short LibrariesManager::getBehaviourType(const std::string &l,
-- 
2.43.0

